lrpar/src/lib/parser.rs

```diff
@@ -643,30 +608,53 @@ pub(crate) mod test {
             .collect();
         let lexer_rules = small_lexer(lexs, rule_ids);
         let lexemes = small_lex(lexer_rules, input);
+        let mut lexer = SmallLexer{lexemes, i: 0};
         let costs_tidx = costs
             .iter()
             .map(|(k, v)| (grm.token_idx(k).unwrap(), v))
             .collect::<HashMap<_, _>>();
-        let pr = parse_rcvry(
-            rcvry_kind,
-            &grm,
-            move |tidx| **costs_tidx.get(&tidx).unwrap_or(&&1),
+        let r = match RTParserBuilder::new(&grm,
             &sgraph,
-            &stable,
-            &lexemes
-        );
-        (grm, pr)
+            &stable)
+            .recoverer(rcvry_kind)
+            .term_costs(&|tidx| **costs_tidx.get(&tidx).unwrap_or(&&1))
+            .parse(&mut lexer) {
+            Ok(r) => Ok(r),
+            Err(LexParseError::ParseError(r1, r2)) => Err((r1, r2)),
+            _ => unreachable!()
+        };
+        (grm, r)
     }
 
     fn check_parse_output(lexs: &str, grms: &str, input: &str, expected: &str) {
         let (grm, pt) = do_parse(RecoveryKind::MF, lexs, grms, input);
         assert_eq!(expected, pt.unwrap().pp(&grm, &input));
     }
 
-    // small_lexer and small_lex are our highly simplified version of lrlex (allowing us to avoid
-    // having to have lrlex as a dependency of lrpar). The format is the same as lrlex *except*:
+    // SmallLexer is ourhighly simplified version of lrlex (allowing us to avoid having to have
+    // lrlex as a dependency of lrpar). The format is the same as lrlex *except*:
     //   * The initial "%%" isn't needed, and only "'" is valid as a rule name delimiter.
     //   * "Unnamed" rules aren't allowed (e.g. you can't have a rule which discards whitespaces).
+    struct SmallLexer<StorageT> {
+        lexemes: Vec<Lexeme<StorageT>>,
+        i: usize
+    }
+
+    impl<StorageT: Hash + PrimInt + Unsigned> Lexer<StorageT> for SmallLexer<StorageT> {
+        fn next(&mut self) -> Option<Result<Lexeme<StorageT>, LexError>> {
+            let old_i = self.i;
+            if old_i < self.lexemes.len() {
```

Comment:
Why can't we do:
```
if self.i < self.lexemes.len() {
    self.i += 1;
    return Some(Ok(self.lexemes[self.i]));
}
```

Reply:
Because that would skip tokens. The first time you called `next` you'd get the token at position 1 (not, as you want, position 0).

Reply:
Ah whoops, missed the `old_i` in the return again. My bad!